### PR TITLE
[IMP] payment,sale: also confirm quotation after multiple partial payments

### PR DESCRIPTION
**Pull request by hote-odoo**
When creating payment links for the customer, the user can now decide on a partial amount to be paid at the time.

Use case: when the customer cannot pay the SO amount with one payment, e.g. limit on the credit card, the user will now have the possibility to set a partial amount on the payment links.
The order will be automatically confirmed when the amount chosen is equal to the remaining amount to be paid.
Additionally, the customer will be notified by email for each payment done instead of only at the confirmation of the order, which will happens only once the total amount is paid.

Task - 2672713

### CODE REVIEW

File: addons/sale/wizard/payment_link_wizard.py [hunk 1]
```diff
@@ -3,13 +3,43 @@
...
+                           - sum(record.invoice_ids.filtered(lambda x: x.state != 'cancel')
+                                 .mapped('amount_total'))
+                           - record.amount_paid())
+            res.update({
+                'amount': amount,
+                'amount_paid': record.amount_paid(),
+                'amount_max': record.amount_total,
+            })
+
+        return res
```
Comment (Feyensv) [inline addons/sale/wizard/payment_link_wizard.py:35]: Could be handled directly in `sale.order` `_get_default_payment_link_values` method ;).

File: addons/payment/wizards/payment_link_wizard.py [hunk 2]
```diff
@@ -54,13 +54,15 @@ def default_get(self, fields_list):
         default='all',
         required=True,
     )
+    warning_message = fields.Char(compute="_compute_warning_message")
 
-    @api.onchange('amount', 'description')
-    def _onchange_amount(self):
+    @api.depends('amount', 'amount_max', 'description')
+    def _compute_warning_message(self):
```
Comment (chevalierv) [inline addons/payment/wizards/payment_link_wizard.py:60]: Well, it works, but I'd instead use `warning_message`  as it is, a warning message :) 

May I suggest keeping the `_on_change_amount`, but that'll set a new variable `is_amount_allowed` to ` false`? You can create a generic warning message in the template that is hidden if the amount is allowed :)

Comment (Feyensv) [reply addons/payment/wizards/payment_link_wizard.py:60]: Indeed, we are somehow mixing warning logic and link validity, which could be two separate concepts in some cases (not the case for now i think, but we could consider having a warning message that still allows the user to copy the link).

File: addons/sale/wizard/payment_link_wizard.py (hunk 1 above)
Comment (hote-odoo) [reply addons/sale/wizard/payment_link_wizard.py]: thanks !

**Review by Demesmaeker — CHANGES_REQUESTED** [review]
Hello :duck: 

The name of this PR and commit should be updated to show that `payment` is impacted IMO
Also, the name is quite misleading, at first, I thought a partial payment would confirm the quotation in any case... Not just when the total paid (from the partial tx) was the quotation amount. :thinking: Not very important, but if you think of something else to name this task, do not hesitate.

Also, some of these changes are only done for the `sale` part, and not in the `account_payment`. And some that should be taken from there too. Both payment link wizards should have the same comportment, IMO. Shouldn't the customers from invoice also get a message? And then, why not also use the ´partial´ banner, as it's done in the invoices? :thinking:

File: addons/payment/wizards/payment_link_wizard_views.xml [hunk 3]
```diff
@@ -23,15 +23,28 @@
...
                         <field name="payment_provider_selection"
                             attrs="{'invisible':[('has_multiple_providers', '=', False)]}"/>
                     </group>
                 </group>
+                <div class="alert alert-warning text-center fw-bold" role="alert" attrs="{'invisible': [('is_amount_allowed', '=', True)]}">
+                    <p attrs="{'invisible': [('amount_max', '!=', 0)]}">
+                        There is nothing to be paid.
+                    </p>
+                    <p attrs="{'invisible': ['|', ('amount', '&lt;', 0), ('amount_max', '=', 0)]}">
+                        Please set an amount lower than <field name="amount_max" readonly="1" widget="monetary"/>.
```
[the remark this answers is not in the document]
Comment (hote-odoo) [reply addons/payment/wizards/payment_link_wizard_views.xml:37]: Then the paragraph below is triggered and "Please set a positive amount." is shown

[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment/wizards/payment_link_wizard_views.xml:37]: You sure of that? :innocent:

File: addons/payment/wizards/payment_link_wizard_views.xml [hunk 4]
```diff
@@ -23,15 +23,28 @@
...
+                    </p>
+                    <p attrs="{'invisible': ['|', ('amount', '>=', 0), ('amount_max', '=', 0)]}">
```
Comment (Demesmaeker) [inline addons/payment/wizards/payment_link_wizard_views.xml:39]: Pretty sure if `amount == 0` then this is invisible =P

File: addons/payment/wizards/payment_link_wizard.py [hunk 5]
```diff
@@ -54,13 +53,11 @@ def default_get(self, fields_list):
...
     )
+    is_amount_allowed = fields.Boolean(compute="_compute_is_amount_allowed")
 
-    @api.onchange('amount', 'description')
-    def _onchange_amount(self):
-        if float_compare(self.amount_max, self.amount, precision_rounding=self.currency_id.rounding or 0.01) == -1:
-            raise ValidationError(_("Please set an amount smaller than %s.", self.amount_max))
-        if self.amount <= 0:
-            raise ValidationError(_("The value of the payment amount must be positive."))
+    @api.depends('amount', 'description')
```
Comment (Demesmaeker) [inline addons/payment/wizards/payment_link_wizard.py:58]: Why the dependency on the `description`?

File: addons/sale/models/payment_transaction.py [hunk 6]
```diff
@@ -49,54 +49,45 @@ def _set_pending(self, state_message=None, **kwargs):
...
+        # for the orders that has been confirmed send the confirmation mail
         return confirmed_orders
 
     def _set_authorized(self, state_message=None, **kwargs):
         """ Override of payment to confirm the quotations automatically. """
         super()._set_authorized(state_message=state_message, **kwargs)
-        confirmed_orders = self._check_amount_and_confirm_order()
-        confirmed_orders._send_order_confirmation_mail()
+        confirmed_orders = self._post_process_payment_and_update_sale_order()
+        confirmed_orders._send_order_payment_done_mail()
```
Comment (Demesmaeker) [inline addons/sale/models/payment_transaction.py]: This was renamed! Not working anymore

Comment (Demesmaeker) [inline addons/sale/models/payment_transaction.py]: Not only `confirmed_orders`, all the orders will get a mail now

File: addons/sale/models/payment_transaction.py [hunk 7]
```diff
@@ -49,54 +49,45 @@ def _set_pending(self, state_message=None, **kwargs):
...
-                                'so_amount': format_amount(
-                                    quotation.env, quotation.amount_total, quotation.currency_id
-                                ),
-                                'tx_amount': format_amount(tx.env, tx.amount, tx.currency_id),
-                            },
-                        )
+                    elif comparison == 1:
+                        # some amount remain to be paid, inform the customer of the payment
+                        # nevertheless
+                        quotation._send_order_payment_executed_mail()
```
Comment (Demesmaeker) [inline addons/sale/models/payment_transaction.py:82]: No, this is done after, currently L90 and L108. You don't have to separate `confirmed` orders from other anymore =]

File: addons/sale/models/payment_transaction.py [hunk 8]
```diff
@@ -49,54 +49,45 @@ def _set_pending(self, state_message=None, **kwargs):
...
-            # send order confirmation mail.
-            sales_orders._send_order_confirmation_mail()
+            sales_orders._send_order_payment_executed_mail()
 
         return txs_to_process
 
-    def _check_amount_and_confirm_order(self):
-        """ Confirm the sales order based on the amount of a transaction.
+    def _post_process_payment_and_update_sale_order(self):
+        """ Update the sale order based on the amount of the transaction.
```
Comment (Demesmaeker) [inline addons/sale/models/payment_transaction.py:57]: The payment was already post_process at this point, you only check the amount of the transaction and confirm the order when necessary here =] The rename isn't necessary IMO

File: addons/sale/models/sale_order.py [hunk 9]
```diff
@@ -1505,6 +1523,22 @@ def name_get(self):
...
+        We consider that transactions in the states 'authorized', 'done' are already
+        paid as the provider side has already given approbation.
+
+        :return: the amount already paid of the SO
+        :rtype: float
+        """
+        for order in self:
+            order.amount_paid = sum(
+                tx.amount for tx in order.transaction_ids if tx.state in ('authorized', 'done')
+            )
```
Comment (Demesmaeker) [inline addons/sale/models/sale_order.py:1541]: Could you move that in the compute methods instead of the business ones?

File: addons/sale/models/sale_order.py [hunk 10]
```diff
@@ -1505,6 +1523,22 @@ def name_get(self):
...
+    def _compute_amount_paid(self):
+        """
+        Sum of the amount paid through all transactions for this SO.
+
+        For SO that are paid in partial amounts returns the amount paid at the moment.
+        We consider that transactions in the states 'authorized', 'done' are already
+        paid as the provider side has already given approbation.
+
+        :return: the amount already paid of the SO
+        :rtype: float
```
Comment (Demesmaeker) [inline addons/sale/models/sale_order.py:1536]: This doesn't return anything

File: addons/sale/wizard/payment_link_wizard.py [hunk 11]
```diff
@@ -3,13 +3,30 @@
...
 
 
 class PaymentLinkWizard(models.TransientModel):
     _inherit = 'payment.link.wizard'
     _description = 'Generate Sales Payment Link'
 
+    amount_paid = fields.Monetary(
+        currency_field='currency_id',
+        help="Amount partially paid on this order.",
+        string="Already Paid"
```
Comment (Demesmaeker) [inline addons/sale/wizard/payment_link_wizard.py:16]:
```suggestion
        string="Already Paid",
```

File: addons/sale/models/sale_order.py [hunk 12]
```diff
@@ -1407,12 +1422,15 @@ def show_line(line):
...
         self.ensure_one()
+        amount = self.amount_total - self.amount_paid
         return {
             'description': self.name,
-            'amount': self.amount_total - sum(self.invoice_ids.filtered(lambda x: x.state != 'cancel' and x.invoice_line_ids.sale_line_ids.order_id == self).mapped('amount_total')),
             'currency_id': self.currency_id.id,
             'partner_id': self.partner_id.id,
-            'amount_max': self.amount_total,
+            'amount': amount,
+            'amount_max': amount,
```
Comment (Demesmaeker) [inline addons/sale/models/sale_order.py:1431]: Both the same?

Comment (hote-odoo) [reply addons/sale/models/sale_order.py:1431]: Yep, do not allow an amount bigger that what remains to be paid and let the initial amount to be equal to the maximal amount that can be paid

File: addons/sale/wizard/payment_link_wizard.py [hunk 13]
```diff
@@ -3,13 +3,29 @@
...
+    show_confirmation_message = fields.Boolean(compute="_compute_show_confirmation_message")
+    is_quotation = fields.Boolean()
+
+    @api.depends('amount')
+    def _compute_show_confirmation_message(self):
+        self.show_confirmation_message = (
+            self.amount_max != 0 and
+            self.amount == self.amount_max and
+            self.res_model == 'sale.order' and
+            self.is_quotation
```
Comment (Demesmaeker) [inline addons/sale/wizard/payment_link_wizard.py:26]:
```suggestion
            self.amount_max and self.amount == self.amount_max and self.is_quotation
```
Why check the res_model here?

**Review by Demesmaeker — CHANGES_REQUESTED** [review]
Hey there

This might be done later, but adding the ribbon when it's paid/partially paid/... on the SO as it's done on the invoice would be nice too IMO

cc: @dalaOdoo

File: addons/sale/wizard/payment_link_wizard.py (hunk 13 above)
Comment (Demesmaeker) [reply addons/sale/wizard/payment_link_wizard.py:26]: To be clearer: when setting `is_quotation`, put the check there to be sure you are not on a `account.move`

File: addons/sale/wizard/payment_link_wizard_views.xml [hunk 14]
```diff
@@ -11,4 +11,25 @@
...
+        <field name="model">payment.link.wizard</field>
+        <field name="inherit_id" ref="payment.payment_link_wizard_view_form"/>
+        <field name="arch" type="xml">
+            <field name="amount" position="after">
+                <field name="show_confirmation_message" invisible="1"/>
+                <field name="amount_paid" readonly="1" attrs="{ 'invisible': [('amount_paid', '&lt;=', 0)]}"/>
+            </field>
+            <div class="alert alert-warning text-center fw-bold"
+                 attrs="{'invisible': [('is_amount_allowed', '=', True)]}"
+                 position="after">
```
Comment (Feyensv) [inline addons/sale/wizard/payment_link_wizard_views.xml:25]: Please add a name in the targeted div and use the name to target it.
It's confusing here.

File: addons/website_sale/models/sale_order.py [hunk 15]
```diff
@@ -442,7 +442,7 @@ def action_confirm(self):
         res = super().action_confirm()
         for order in self:
             if not order.transaction_ids and not order.amount_total and self._context.get('send_email'):
-                order._send_order_confirmation_mail()
+                order._send_payment_succeeded_for_order_mail()
```
Comment (Feyensv) [inline addons/website_sale/models/sale_order.py:445]: This is not related to a payment at all.
To investigate, but either this override can be dropped, or we have to keep the `_send_order_confirmation_email` logic.

File: addons/sale/models/sale_order.py [hunk 16]
```diff
@@ -905,7 +914,11 @@ def _action_confirm(self):
                 if not order.analytic_account_id:
                     order._create_analytic_account()
 
-    def _send_order_confirmation_mail(self):
+    def _send_payment_succeeded_for_order_mail(self):
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:917]: Confusing logic mixing order confirmation mail and payment succeeded.
I think the logic should be kept separated one way or another :thinking:.

File: addons/sale/controllers/portal.py [hunk 17]
```diff
@@ -244,7 +244,7 @@ def portal_quote_accept(self, order_id, access_token=None, name=None, signature=
 
         if not order_sudo._has_to_be_paid():
             order_sudo.action_confirm()
-            order_sudo._send_order_confirmation_mail()
+            order_sudo._send_payment_succeeded_for_order_mail()
```
Comment (Feyensv) [inline addons/sale/controllers/portal.py:247]: There is no payment logic here, we only confirmed a SO by signing it.
We get the impression that you are sending a 'payment succeeded' mail when no payment happened at all.

Okay, the logic behind it will send the right mail, but it's confusing to read and understand.

**Review by Feyensv — COMMENTED** [review]
I believe `_send_confirmation_mail` still has sense and must be kept.
The method to send the payment succeeded mail should be separate, unless you find a really clean and clear way to share the logic in one method.

File: addons/payment/wizards/payment_link_wizard.py [hunk 18]
```diff
@@ -54,13 +53,14 @@ def default_get(self, fields_list):
...
-
-    @api.onchange('amount', 'description')
-    def _onchange_amount(self):
-        if float_compare(self.amount_max, self.amount, precision_rounding=self.currency_id.rounding or 0.01) == -1:
-            raise ValidationError(_("Please set an amount smaller than %s.", self.amount_max))
-        if self.amount <= 0:
-            raise ValidationError(_("The value of the payment amount must be positive."))
+    is_amount_allowed = fields.Boolean(compute="_compute_is_amount_allowed")
+
+    @api.depends('amount', 'description')
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/payment/wizards/payment_link_wizard.py]: Yes :D

File: addons/sale/models/sale_order.py [hunk 19]
```diff
@@ -286,6 +287,7 @@ def _rec_names_search(self):
     tax_totals = fields.Binary(compute='_compute_tax_totals', exportable=False)
     terms_type = fields.Selection(related='company_id.terms_type')
     type_name = fields.Char(string="Type Name", compute='_compute_type_name')
+    amount_paid = fields.Float(compute='_compute_amount_paid')
```
Comment (Demesmaeker) [inline addons/sale/models/sale_order.py:290]: Duplicate code

File: addons/sale/models/sale_order.py [hunk 20]
```diff
@@ -645,6 +655,20 @@ def _compute_access_url(self):
...
+
+        For SO that are paid in partial amounts returns the amount paid at the moment.
+        We consider that transactions in the states 'authorized', 'done' are already
+        paid as the provider side has already given approbation.
+        """
+        for order in self:
+            order.amount_paid = sum(
+                tx.amount for tx in order.transaction_ids if tx.state in ('authorized', 'done')
+            )
+
```
Comment (Demesmaeker) [inline addons/sale/models/sale_order.py:671]: Duplicate code

File: addons/sale/models/sale_order.py [hunk 21]
```diff
@@ -906,6 +938,18 @@ def _action_confirm(self):
                     order._create_analytic_account()
 
     def _send_order_confirmation_mail(self):
+        """
+        Send a mail to the customer to notify him that its order has been confirmend.
+
+        :return: None
+        """
+        self._send_payment_succeeded_for_order_mail()
```
Comment (Demesmaeker) [inline addons/sale/models/sale_order.py:946]: Could you separate the two methods to not make one call the other anymore?

**Review by Demesmaeker — COMMENTED** [review]
Heyho :duck:

File: addons/sale/models/sale_order.py (hunk 21 above)
Comment (hote-odoo) [reply addons/sale/models/sale_order.py:946]: I was doing that at the beginning but you will have the same code on both, the only difference is the template that is called. I would prefer if we do not have two methods that do the same stuff.

Comment (Demesmaeker) [reply addons/sale/models/sale_order.py:931]: It won't be the same code, one will have the confirmation template, and not the other. I don't see the problem here

File: addons/sale/models/sale_order.py [hunk 22]
```diff
@@ -906,22 +915,50 @@ def _action_confirm(self):
...
+            sale_order._send_order_notification_mail(mail_template)
+
+    def _send_order_notification_mail(self, mail_template):
+        """ Send a mail to the customer
+
+        :mail_template: a mail template record to generate the mail
+        :return: None
+        """
+
+        # common code using `mail_template`
```
Comment (Demesmaeker) [inline addons/sale/models/sale_order.py:945]: You can get rid of this comment, it was just to explain what I had in mind =]

File: addons/sale/models/sale_order.py [hunk 23]
```diff
@@ -906,22 +915,50 @@ def _action_confirm(self):
...
+        for sale_order in self:
+            mail_template = self.env.ref(
+                'sale.mail_template_sale_payment_executed', raise_if_not_found=False
+            )
+            sale_order._send_order_notification_mail(mail_template)
+
+    def _send_order_notification_mail(self, mail_template):
+        """ Send a mail to the customer
+
+        :mail_template: a mail template record to generate the mail
```
Comment (Demesmaeker) [inline addons/sale/models/sale_order.py:941]:
```suggestion
        :param mail.template mail_template: the template used to generate the mail
```

File: addons/sale/models/sale_order.py [hunk 24]
```diff
@@ -906,22 +915,50 @@ def _action_confirm(self):
...
+        """
+        for sale_order in self:
+            mail_template = self.env.ref(
+                'sale.mail_template_sale_payment_executed', raise_if_not_found=False
+            )
+            sale_order._send_order_notification_mail(mail_template)
+
+    def _send_order_notification_mail(self, mail_template):
+        """ Send a mail to the customer
+
```
Comment (Demesmaeker) [inline addons/sale/models/sale_order.py:940]:
```suggestion

        Note: self.ensure_one()

```

**Review by Demesmaeker — COMMENTED** [review]
Almost there =]

File: addons/sale/models/sale_order.py [hunk 25]
```diff
@@ -906,22 +915,50 @@ def _action_confirm(self):
...
-                email_layout_xmlid='mail.mail_notification_layout_with_responsible_signature',
-                subtype_xmlid='mail.mt_comment',
-            )
+        if not mail_template:
+            continue
+        self.with_context(force_send=True).message_post_with_source(
+            mail_template,
+            email_layout_xmlid='mail.mail_notification_layout_with_responsible_signature',
+            subtype_xmlid='mail.mt_comment',
+        )
```
Comment (Demesmaeker) [inline addons/sale/models/sale_order.py:960]:
```suggestion
        if mail_template:
            self.with_context(force_send=True).message_post_with_source(
                mail_template,
                email_layout_xmlid='mail.mail_notification_layout_with_responsible_signature',
                subtype_xmlid='mail.mt_comment',
            )
```

File: addons/sale/models/sale_order.py [hunk 26]
```diff
@@ -906,23 +915,50 @@ def _action_confirm(self):
...
+    def _send_order_notification_mail(self, mail_template):
+        """ Send a mail to the customer
+
+        Note: self.ensure_one()
+
+        :param mail.template mail_template: the template used to generate the mail
+        :return: None
+        """
         if not self:
             return
```
Comment (Demesmaeker) [inline addons/sale/models/sale_order.py:947]:
```suggestion
        self.ensure_one()
        
        if not mail_template:
            return
```

In fact, this could get rid of the last if easily =] Sorry, should have seen that sooner

File: addons/sale/models/sale_order.py [hunk 27]
```diff
@@ -906,23 +915,50 @@ def _action_confirm(self):
...
-                continue
-            sale_order.with_context(force_send=True).message_post_with_source(
+        if mail_template:
+            self.with_context(force_send=True).message_post_with_source(
                 mail_template,
                 email_layout_xmlid='mail.mail_notification_layout_with_responsible_signature',
                 subtype_xmlid='mail.mt_comment',
             )
 
+
```
Comment (Demesmaeker) [inline addons/sale/models/sale_order.py:961]:
```suggestion
```

**Review by Demesmaeker — APPROVED** [review]
lgtm after the 2 small modifications ^^

File: addons/sale/models/sale_order.py (hunk 26 above)
Comment (hote-odoo) [reply addons/sale/models/sale_order.py:947]: yep, whenever called self is not empty :)

File: addons/sale/wizard/payment_link_wizard_views.xml [hunk 28]
```diff
@@ -11,4 +11,23 @@
         <field name="binding_view_types">form</field>
     </record>
 
+    <record id="payment_link_wizard_view_form_inherit" model="ir.ui.view">
```
Comment (Feyensv) [inline addons/sale/wizard/payment_link_wizard_views.xml:14]:
```suggestion
    <record id="payment_link_wizard_view_form" model="ir.ui.view">
```

File: addons/payment/wizards/payment_link_wizard_views.xml [hunk 29]
```diff
@@ -23,15 +23,28 @@
                         <field name="has_multiple_providers" invisible="1"/>
                         <field name="description"/>
                         <field name="amount"/>
+                        <field name="is_amount_allowed" invisible="1"/>
                         <field name="currency_id" invisible="1"/>
                         <field name="payment_provider_selection"
                             attrs="{'invisible':[('has_multiple_providers', '=', False)]}"/>
                     </group>
                 </group>
+                <div name="payment-link-warning-information" class="alert alert-warning text-center fw-bold" role="alert" attrs="{'invisible': [('is_amount_allowed', '=', True)]}">
```
Comment (Feyensv) [inline addons/payment/wizards/payment_link_wizard_views.xml:32]:
```suggestion
                <div name="payment_link_warning_information" class="alert alert-warning text-center fw-bold" role="alert" attrs="{'invisible': [('is_amount_allowed', '=', True)]}">
```
No - in code, prefer _

File: addons/payment/wizards/payment_link_wizard_views.xml [hunk 30]
```diff
@@ -23,15 +23,28 @@
...
+                    <p attrs="{'invisible': [('amount_max', '!=', 0)]}">
+                        There is nothing to be paid.
+                    </p>
+                    <p attrs="{'invisible': ['|', ('amount', '&lt;=', 0), ('amount_max', '=', 0)]}">
+                        Please set an amount lower than <field name="amount_max" readonly="1" widget="monetary"/>.
```
Comment (Feyensv) [inline addons/payment/wizards/payment_link_wizard_views.xml:37]: Probably not correctly rendered in some languages.
Can't we use a generic field to host the error message, and show the whole diff only if the field is not empty ?

File: addons/payment/wizards/payment_link_wizard.py [hunk 31]
```diff
@@ -54,13 +53,11 @@ def default_get(self, fields_list):
...
     )
+    is_amount_allowed = fields.Boolean(compute="_compute_is_amount_allowed")
 
-    @api.onchange('amount', 'description')
-    def _onchange_amount(self):
-        if float_compare(self.amount_max, self.amount, precision_rounding=self.currency_id.rounding or 0.01) == -1:
-            raise ValidationError(_("Please set an amount smaller than %s.", self.amount_max))
-        if self.amount <= 0:
-            raise ValidationError(_("The value of the payment amount must be positive."))
+    @api.depends('amount')
```
Comment (Feyensv) [inline addons/payment/wizards/payment_link_wizard.py:58]:
```suggestion
    @api.depends('amount', 'amount_max')
```

File: addons/sale/wizard/payment_link_wizard.py [hunk 32]
```diff
@@ -3,13 +3,26 @@
...
-from odoo import api, models
+from odoo import api, fields, models
 
 
 class PaymentLinkWizard(models.TransientModel):
     _inherit = 'payment.link.wizard'
     _description = 'Generate Sales Payment Link'
 
+    amount_paid = fields.Monetary(
+        currency_field='currency_id',
```
Comment (Feyensv) [inline addons/sale/wizard/payment_link_wizard.py:14]:
```suggestion
```
No need, it's the default value for Monetary fields

File: addons/sale/wizard/payment_link_wizard.py [hunk 33]
```diff
@@ -3,13 +3,26 @@
...
+        string="Already Paid",
+    )
+    show_confirmation_message = fields.Boolean(compute="_compute_show_confirmation_message")
+    is_quotation = fields.Boolean()
```
Comment (Feyensv) [inline addons/sale/wizard/payment_link_wizard.py:18]: Is this really necessary ?
We have the res_model, res_id, there is no need for a new field here.

File: addons/sale/models/sale_order.py [hunk 34]
```diff
@@ -906,22 +915,48 @@ def _action_confirm(self):
...
+    def _send_payment_succeeded_for_order_mail(self):
+        """ Send a mail to the customer each time a payment is executed.
+
+        :return: None
+        """
+        for sale_order in self:
+            mail_template = self.env.ref(
+                'sale.mail_template_sale_payment_executed', raise_if_not_found=False
+            )
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:935]: The `ref` is independent of the orders, must be moved outside the loop (ref calls cost one SQL query, you are making len(self) calls for 'nothing' here)

File: addons/sale/data/mail_template_data.xml [hunk 35]
```diff
@@ -240,6 +240,59 @@
...
+        A payment with reference
+        <span style="font-weight:bold;" t-out="transaction.reference or ''">SOOO49</span>
+        amounting
+        <span style="font-weight:bold;" t-out="format_amount(transaction.amount, object.currency_id) or ''">$ 10.00</span>
+        for your order
+        <span style="font-weight:bold;" t-out="object.name or ''">S00049</span>
+        <t t-if="transaction and transaction.state == 'pending'">
+            is pending.
+            <br/>
+            <t t-if="object.amount_paid + transaction.amount >= object.amount_total and object.state in ('draft', 'sent')">
```
Comment (Feyensv) [inline addons/sale/data/mail_template_data.xml:265]: currency compare_amounts ?

**Review by Feyensv — COMMENTED** [review]
:duck:

File: addons/payment/wizards/payment_link_wizard.py [hunk 36]
```diff
@@ -54,13 +53,18 @@ def default_get(self, fields_list):
...
+    @api.depends('amount', 'amount_max')
+    def _compute_warning_message(self):
+        self.warning_message = ''
+        for link in self:
+            if link.amount_max <= 0:
+                link.warning_message = _("There is nothing to be paid.")
+            if link.amount <= 0:
+                link.warning_message = _("Please set a positive amount.")
+            if link.amount > link.amount_max:
+                link.warning_message = _("Please set an amount lower than %s.", link.amount_max)
```
Comment (Feyensv) [inline addons/payment/wizards/payment_link_wizard.py:67]: Either you must use `+-` and new lines, or you should use `elif`'s here

File: addons/sale/models/sale_order.py [hunk 37]
```diff
@@ -908,22 +917,50 @@ def _action_confirm(self):
...
+            sale_order._send_order_notification_mail(mail_template)
+
+    def _send_payment_succeeded_for_order_mail(self):
+        """ Send a mail to the customer each time a payment is executed.
+
+        :return: None
+        """
+        mail_template = self.env.ref(
+                'sale.mail_template_sale_payment_executed', raise_if_not_found=False
+            )
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:938]: wrong indentation

File: addons/sale/models/sale_order.py [hunk 38]
```diff
@@ -908,22 +917,50 @@ def _action_confirm(self):
...
     def _send_order_confirmation_mail(self):
-        if not self:
+        """ Send a mail to the SO customer to inform them that their order has been confirmed.
+
+        :return: None
+        """
+        if not self:  # This method can be called with an empty recordset and
+                      # calling `_get_confirmation_template` will crash.
+            return
+        mail_template = self._get_confirmation_template()
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:927]: Wrong, `_get_confirmation_template` is `ensure_one()`, it must stay in the loop here.

File: addons/sale/wizard/payment_link_wizard.py [hunk 39]
```diff
@@ -3,13 +3,28 @@
...
 class PaymentLinkWizard(models.TransientModel):
     _inherit = 'payment.link.wizard'
     _description = 'Generate Sales Payment Link'
 
+    amount_paid = fields.Monetary(string="Already Paid")
+    show_confirmation_message = fields.Boolean(compute="_compute_show_confirmation_message")
+
+    @api.depends('amount')
+    def _compute_show_confirmation_message(self):
+
```
Comment (Feyensv) [inline addons/sale/wizard/payment_link_wizard.py:18]:
```suggestion
```

File: addons/sale/models/sale_order.py [hunk 40]
```diff
@@ -908,22 +917,47 @@ def _action_confirm(self):
...
+            mail_template = self._get_confirmation_template()
+            sale_order._send_order_notification_mail(mail_template)
+
+    def _send_payment_succeeded_for_order_mail(self):
+        """ Send a mail to the customer each time a payment is executed.
+
+        :return: None
+        """
+        mail_template = self.env.ref(
+                'sale.mail_template_sale_payment_executed', raise_if_not_found=False
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:934]:
```suggestion
            'sale.mail_template_sale_payment_executed', raise_if_not_found=False
```
...

File: addons/sale/wizard/payment_link_wizard_views.xml [hunk 41]
```diff
@@ -11,4 +11,23 @@
...
     </record>
 
+    <record id="payment_link_wizard_view_form" model="ir.ui.view">
+        <field name="name">payment.link.wizard.form</field>
+        <field name="model">payment.link.wizard</field>
+        <field name="inherit_id" ref="payment.payment_link_wizard_view_form"/>
+        <field name="arch" type="xml">
+            <field name="amount" position="after">
+                <field name="show_confirmation_message" invisible="1"/>
+                <field name="amount_paid" attrs="{ 'invisible': [('amount_paid', '&lt;=', 0)]}"/>
```
Comment (Feyensv) [inline addons/sale/wizard/payment_link_wizard_views.xml:21]:
```suggestion
                <field name="amount_paid" attrs="{'invisible': [('amount_paid', '&lt;=', 0)]}"/>
```
already mentioned in latest review

**Review by Feyensv — COMMENTED** [review]
Pending payments are sending the confirmation mail when they should send the payment mail, cf fixup commit I pushed

**Comment by Feyensv** [comment]
@robodoo r+

**Comment by Feyensv** [comment]
@robodoo r+

File: addons/sale/models/sale_order.py [hunk 42]
```diff
@@ -908,22 +917,47 @@ def _action_confirm(self):
                     order._create_analytic_account()
 
     def _send_order_confirmation_mail(self):
-        if not self:
+        """ Send a mail to the SO customer to inform them that their order has been confirmed.
+
+        :return: None
+        """
+        for order in self:
+            mail_template = self._get_confirmation_template()
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:925]: order._get_confirmation_template

:eyes:
